ui: e2e tests, improved cypress test for statement page

nanduu04 opened the pull request:
Previously, only visual regression tests were written

Added cypress-testing-library which provides better API to write
cypress-tests which allows us to get rid of CSS-selectors

Cypress test for statement page was written using the cypress-testing-library

[CRDB-2392](https://cockroachlabs.atlassian.net/browse/CRDB-2386?atlOrigin=eyJpIjoiM2JiM2YzNDFmMzdkNDYwZGI2NTNiMGQ2MjUxZDM0YjciLCJwIjoiaiJ9)

[CRDB-2387](https://cockroachlabs.atlassian.net/browse/CRDB-2387?atlOrigin=eyJpIjoiNzBhYmUzMjA5NGY0NDM3OGEyNzM0MGI5NzJiM2M5ZmIiLCJwIjoiaiJ9)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61175)
<!-- Reviewable:end -->

Commit:
ui: e2e tests, improved cypress test for statement page

Previously, only visual regression tests were written

Added cypress-testing-library which provides better API to write
cypress-tests which allows us to get rid of CSS-selectors. Bumped
the typescript version to 3.8.0 to make the
 cypress-testing-library work.

Cypress test for statement page was written using the cypress-testing-library

Release note: None

dhartunian reviewed (commented):
> (not ready for review yet)

I think it might be better to get this PR out of draft and get it merged. For this pass, let's focus on just the Statements tests and make sure those are in good shape. No need to also complete the Transactions, Sessions, etc. Once you get this one merged, the rest will be easier since we'll have completed the feedback round.

> Reviewed 5 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-:-MWyFJdECvufX_g5wSGw:bkalfyy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nanduu04)

___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFQT52tnPpw9PbaVz:-MWyFQT52tnPpw9PbaW-:b9se215) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L39)):*
> ```XML
>     cy.get(".ant-modal-content").contains("Activate statement diagnostics");
>     cy.log("Click the button for the diagonistics");
>     cy.get("div> div > div > div > div > button").eq(2).click();
> ```

You should replace selectors like this with the `cypress-testing-library`. That will let you target buttons based on what their text is. From this selector you can't tell what the button actually is.
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFc5f2ycV7nmbcliR:-MWyFc5f2ycV7nmbcliS:b-rq5x5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L43)):*
> ```XML
>       'cockroach sql --insecure --execute="insert into test values (1)";'
>     );
>     cy.reload();
> ```

After you reload, the test should check for something. Right now it doesn't check anything.
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFjgsAPu5NiuqLypv:-MWyFjgsAPu5NiuqLypw:bmn3c15) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L50)):*
> ```XML
>   it("renders default view", () => {
>     cy.visit("#/statements");
>     cy.get(".ant-input").type("insert into test").get(".ant-btn").click();
> ```

Each test should be independent of others. You are doing an `insert` in the previous test and then expecting it in this one. You should assume every single test can be executed independently and they should all be able to run successfully either on their own or all together (in any order).
___
*[pkg/ui/cypress/support/index.ts, line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyG2Vk7db9heoCtFf9:-MWyG2Vk7db9heoCtFfA:b8525vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/support/index.ts#L13)):*
> ```XML
> import "./commands";
> Cypress.on("uncaught:exception", (err, runnable) => {
>   // returning false here prevents Cypress from
> ```

Why was this necessary to add? Are you seeing uncaught exceptions that make tests fail unnecessarily? If so, maybe add some more information in the comment about why it's necessary.
___
*[pkg/ui/src/components/modal/modal.tsx, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyGBtT4oO56lDqp-JY:-MWyGBtT4oO56lDqp-JZ:bz8kccf) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/src/components/modal/modal.tsx#L58)):*
> ```XML
>           intent="primary"
>           key="okButton"
>           data-testid="Activate"
> ```

Hopefully the new testing library makes these `data-testid` attributes obsolete


<!-- Sent from Reviewable.io -->

nanduu04 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @nanduu04)

___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFQT52tnPpw9PbaVz:-MX5u_-I-gzGO0Gzu4wW:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L39)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

You should replace selectors like this with the `cypress-testing-library`. That will let you target buttons based on what their text is. From this selector you can't tell what the button actually is.
</blockquote></details>

Done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFc5f2ycV7nmbcliR:-MX5uaPp4ogbIljuIiEM:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L43)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

After you reload, the test should check for something. Right now it doesn't check anything.
</blockquote></details>

done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFjgsAPu5NiuqLypv:-MX5ublz9cOlutNVUtqX:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L50)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Each test should be independent of others. You are doing an `insert` in the previous test and then expecting it in this one. You should assume every single test can be executed independently and they should all be able to run successfully either on their own or all together (in any order).
</blockquote></details>

done
___
*[pkg/ui/src/components/modal/modal.tsx, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyGBtT4oO56lDqp-JY:-MX5ugkQ0dyfN9v5cd7_:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/src/components/modal/modal.tsx#L58)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Hopefully the new testing library makes these `data-testid` attributes obsolete
</blockquote></details>

removed


<!-- Sent from Reviewable.io -->

dhartunian reviewed (changes requested):
> Reviewed 4 of 14 files at r1, 3 of 4 files at r3, 1 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @nanduu04)

___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdh8lo9kj4kMRIfcjN:-MXdh8lo9kj4kMRIfcjO:bx0gde5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L23)):*
> ```XML
> describe("Statements - Check whether the written queries match on the statement page", () => {
>   it("Writes some queries and checks them", () => {
>     cy.exec("cockroach workload init movr --drop");
> ```

This is unnecessary I think, you can remove.
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdhMpu8Vw62mIy6Pi-:-MXdhMpu8Vw62mIy6Pi0:b-97goyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L35)):*
> ```XML
> });
> 
> describe("Statements - Check whether diagnostics have been downloaded", () => {
> ```

Let's change this test to "can successfully activate diagnostics" instead of testing the full download
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdhCTw580HHYgsSir8:-MXdhCTw580HHYgsSir9:bx7skv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L36)):*
> ```XML
> 
> describe("Statements - Check whether diagnostics have been downloaded", () => {
>   it("renders default view", () => {
> ```

Every test should satrt with `cy.visit` otherwise it can't be run on its own.
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdhXUR4SyMtXWIixkv:-MXdhXUR4SyMtXWIixkw:b-xky8pl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L42)):*
> ```XML
>     cy.get("td > div > button").eq(-1).should("contain", "Activate").click();
>     cy.get(".ant-modal-content").contains("Activate statement diagnostics");
>     cy.log("Click the button for the diagonistics");
> ```

spelling: diagnostics
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 55 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdhbZx76IGcCR_x87_:-MXdhbZx76IGcCR_x87a:b-qnbgy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L55)):*
> ```XML
>     cy.get("tr").find("td > div > div").eq(-1).click();
>     let statementBundleId = "";
>     cy.get("td > div")
> ```

Remove the clicking on the download for now and leave a `// TODO` comment for us to add in the future


<!-- Sent from Reviewable.io -->

nanduu04 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @nanduu04)

___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFQT52tnPpw9PbaVz:-MXhM9lO5aum8d1cOLBS:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L39)):*
<details><summary><i>Previously, nanduu04 (Nandu Pokhrel) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFc5f2ycV7nmbcliR:-MXhMArc4fbmUScb--T0:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L43)):*
<details><summary><i>Previously, nanduu04 (Nandu Pokhrel) wrote…</i></summary><blockquote>

done
</blockquote></details>

Done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MWyFjgsAPu5NiuqLypv:-MXhMBii2kMGk8gr4Xs4:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/655e665285ca74df9bebf8542a094d03bce9e8a7/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L50)):*
<details><summary><i>Previously, nanduu04 (Nandu Pokhrel) wrote…</i></summary><blockquote>

done
</blockquote></details>

Done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdh8lo9kj4kMRIfcjN:-MXhM5ecEb8ZQ7gIxzjq:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L23)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This is unnecessary I think, you can remove.
</blockquote></details>

Done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdhMpu8Vw62mIy6Pi-:-MXhM7nM6l-TEtGcFVK_:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L35)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Let's change this test to "can successfully activate diagnostics" instead of testing the full download
</blockquote></details>

Done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdhCTw580HHYgsSir8:-MXhMI6VEa2bewpvUSN3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L36)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Every test should satrt with `cy.visit` otherwise it can't be run on its own.
</blockquote></details>

Done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdhXUR4SyMtXWIixkv:-MXhM8itCKXV4PzRIPwK:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L42)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

spelling: diagnostics
</blockquote></details>

Done
___
*[pkg/ui/cypress/integration/statements/statements.visual.spec.ts, line 55 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61175#-MXdhbZx76IGcCR_x87_:-MXhMCboB7UwBBmOCur4:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d517de6a5245e2d27ecc647ba9a75dc3f6df0bf1/pkg/ui/cypress/integration/statements/statements.visual.spec.ts#L55)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Remove the clicking on the download for now and leave a `// TODO` comment for us to add in the future
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
> Reviewed 2 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @nanduu04)



<!-- Sent from Reviewable.io -->

nanduu04 commented:
bors r+
